### PR TITLE
[Refactor] Classification 9/n

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Continues work on classification refactor https://github.com/Lightning-AI/metrics/issues/1001
Prior work:

* https://github.com/Lightning-AI/metrics/pull/1175
* https://github.com/Lightning-AI/metrics/pull/1167
* https://github.com/Lightning-AI/metrics/pull/1163
* https://github.com/Lightning-AI/metrics/pull/1159
* https://github.com/Lightning-AI/metrics/pull/1151
* https://github.com/Lightning-AI/metrics/pull/1145
* https://github.com/Lightning-AI/metrics/pull/1143
* https://github.com/Lightning-AI/metrics/pull/1054

At this point the core implementations are in place and clean up is beginning. This PR takes care of 3 issues:
1. functional `auc` is deprecated as an metric and moved to utilities as an compute tool instead. Module `AUC` is completly deprecated and will be removed in future.
2. `kl_divergence` is deprecated from classification and moved to regression where it better fits (based on the fact that `target` should be a float tensor for this metric)
3. the default value of `average` is changed from `micro` to `macro`. This is to prevent uses getting the same value from `precision`, `recall` and `f1` in the multiclass/multilabel cases. Essentially the following holds:
   ```python
   from torchmetrics.functional import multiclass_precision, multiclass_recall
   import torch
   p, t = torch.randint(3, (20,)), torch.randint(3, (20,))
   multiclass_precision(p, t, num_classes=3, average="micro") == multiclass_recall(p, t, num_classes=3, average="micro")
   # tensor(True)
   multiclass_precision(p, t, num_classes=3, average="macro") == multiclass_recall(p, t, num_classes=3, average="macro")
   # tensor(False)
   ```
   it is just not wrong to use `micro` in these cases, the underlying statistics just collapse into the same numbers in these cases. This change should hopefully prevent such confusion for users.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/torchmetrics/utilities/compute.py [hunk 1]
```diff
@@ -49,3 +53,63 @@ def _safe_divide(num: Tensor, denom: Tensor) -> Tensor:
     num = num if num.is_floating_point() else num.float()
     denom = denom if denom.is_floating_point() else denom.float()
     return num / denom
+
+
+def _auc_update(x: Tensor, y: Tensor) -> Tuple[Tensor, Tensor]:
```
Comment (justusschock) [inline src/torchmetrics/utilities/compute.py:58]:
```suggestion
def _auc_format_inputs(x: Tensor, y: Tensor) -> Tuple[Tensor, Tensor]:
```

File: src/torchmetrics/utilities/compute.py [hunk 2]
```diff
@@ -49,3 +53,63 @@ def _safe_divide(num: Tensor, denom: Tensor) -> Tensor:
...
+
+    Args:
+        x: x-coordinates, must be either increasing or decreasing
+        y: y-coordinates
+        reorder: if True, will reorder the arrays to make it either increasing or decreasing
+
+    Return:
+        Tensor containing AUC score
+    """
+    x, y = _auc_update(x, y)
```
Comment (justusschock) [inline src/torchmetrics/utilities/compute.py:114]:
```suggestion
    x, y = _auc_format_inputs(x, y)
```

**Review by justusschock — APPROVED** [review]
We probably don't strictly need to follow update-compute in our utils, especially if there is not modular interface so I propose a slightly more descriptive name :)

Other than that: LGTM as always :)
